### PR TITLE
CORTX-33462 - Added Copy Object tests with IAM Policy constrains

### PR DESCRIPTION
**Pull request by ketanarlulkar**
Automated test defined in Table 6: Scenarios for inline IAM user policy for Copy Object of [F-20A: S3 Copy Object API](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/1079640340/TDS+F-20A+S3+Copy+Object+API)

### CODE REVIEW

File: s3tests_boto3/functional/test_iam.py [hunk 1]
```diff
@@ -1591,3 +1593,171 @@ def test_deny_version_id_condition_in_user_policy():
...
+    response = client.put_user_policy(PolicyDocument=policy_document_allow,
+                                      PolicyName='AllowPutPolicy', UserName=get_alt_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    # Copy object user2buck/srcobj to user1buck/dstobj  as user2
+    copy_source = {"Bucket": user2buck, 'Key': 'srcobj'}
+    response = s3_client_alt.copy_object(Bucket=user1buck, CopySource=copy_source, Key='dstobj')
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = client.delete_user_policy(PolicyName='AllowPutPolicy', UserName=get_alt_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = s3_client_alt.head_object(Bucket=user1buck, Key="dstobj")
```
Comment (ravindrachoudhari) [inline s3tests_boto3/functional/test_iam.py:1637]: Should we add similar check for user2buck/srcobj also just to make sure after doing copy object original obj is not affected?

Comment (archanabpatil) [inline s3tests_boto3/functional/test_iam.py:1637]: 1. by default you have this permission on destination bucket?
2. Do you want to do GetObject or headObject? (TDS says GetObject)

File: s3tests_boto3/functional/test_iam.py [hunk 2]
```diff
@@ -1591,3 +1593,171 @@ def test_deny_version_id_condition_in_user_policy():
...
+    s3_client_alt = get_alt_client()
+    user2buck = get_new_bucket(client=s3_client_alt)
+    s3_client_alt.put_object(Bucket=user2buck, Key='srcobj', Body='bar')
+    copy_source = {"Bucket": user2buck, 'Key': 'srcobj'}
+    e = assert_raises(ClientError, s3_client_alt.copy_object, Bucket=user1buck,
+                      CopySource=copy_source, Key='dstobj')
+    status, error_code = _get_status_and_error_code(e.response)
+    eq(status, 403)
+    eq(error_code, 'AccessDenied')
+    # Add inline user policy with Allow s3:GetObject on user1buck for user2
```
Comment (archanabpatil) [inline s3tests_boto3/functional/test_iam.py]: this comment should be corrected, we are denying put permission here

File: s3tests_boto3/functional/test_iam.py [hunk 3]
```diff
@@ -1591,3 +1593,171 @@ def test_deny_version_id_condition_in_user_policy():
...
+    client = get_iam_client()
+    # Create bucket user1buck and upload object srcobj as user1
+    s3_client_iam = get_iam_s3client()
+    user1buck = get_new_bucket(client=s3_client_iam)
+    # Create bucket user2buck as user2
+    s3_client_alt = get_alt_client()
+    user2buck = get_new_bucket(client=s3_client_alt)
+    s3_client_alt.put_object(Bucket=user2buck, Key='srcobj', Body='bar')
+    copy_source = {"Bucket": user2buck, 'Key': 'srcobj'}
+    e = assert_raises(ClientError, s3_client_alt.copy_object, Bucket=user1buck,
```
Comment (archanabpatil) [inline s3tests_boto3/functional/test_iam.py:1654]: line 1639-1643 is needed?

File: s3tests_boto3/functional/test_iam.py [hunk 4]
```diff
@@ -1591,3 +1593,171 @@ def test_deny_version_id_condition_in_user_policy():
...
+             "Resource": f"arn:aws:s3:::{user1buck}/*"}}
+    )
+    response = client.put_user_policy(PolicyDocument=policy_document_allow,
+                                      PolicyName='AllowGetPolicy', UserName=get_alt_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    # Copy object user1buck/srcobj to user2buck/dstobj  as user2
+    copy_source = {"Bucket": user1buck, 'Key': 'srcobj'}
+    response = s3_client_alt.copy_object(Bucket=user2buck, CopySource=copy_source, Key='dstobj')
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    # GET Object user1buck/dstobj as user2
```
Comment (archanabpatil) [inline s3tests_boto3/functional/test_iam.py]: user1buck does not have destobj. Please modify this comment and in tds as well

File: s3tests_boto3/functional/test_iam.py [hunk 5]
```diff
@@ -1591,3 +1593,171 @@ def test_deny_version_id_condition_in_user_policy():
...
+         ]}
+    )
+    response = client.put_user_policy(PolicyDocument=policy_document_allow,
+                                      PolicyName='AllowPolicy', UserName=get_main_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    # Copy object user1buck/srcobj to user2buck/dstobj as user3
+    main_s3_client = get_client()
+    copy_source = {"Bucket": user1buck, 'Key': 'srcobj'}
+    main_s3_client.copy_object(Bucket=user2buck, CopySource=copy_source, Key='dstobj')
+    # GET Object user2buck/dstobj as user2
```
Comment (archanabpatil) [inline s3tests_boto3/functional/test_iam.py:1775]: in comment its mentioned , as user2, but you are performing that operation on 3rd user. pls modify the comment  and in tds as well if that (user3)is expected

Comment (archanabpatil) [reply s3tests_boto3/functional/test_iam.py:1775]: we have not permitted user3 for get object on bucket. Pls check and modify appropriately, which user needs to perform get on user2buck/dstobj

**Review by archanabpatil — CHANGES_REQUESTED** [review]

File: s3tests_boto3/functional/test_iam.py (hunk 2 above)
Comment (ketanarlulkar) [reply s3tests_boto3/functional/test_iam.py:1644]: Yes to validate client receives AccessDenied error

File: s3tests_boto3/functional/test_iam.py (hunk 3 above)
Comment (ketanarlulkar) [reply s3tests_boto3/functional/test_iam.py:1654]: Yes to validate client receives AccessDenied error

File: s3tests_boto3/functional/test_iam.py [hunk 6]
```diff
@@ -1591,3 +1593,171 @@ def test_deny_version_id_condition_in_user_policy():
...
+        {"Version": "2012-10-17",
+         "Statement": {
+             "Effect": "Deny",
+             "Action": ["s3:PutObject"],
+             "Resource": f"arn:aws:s3:::{user1buck}/*"}}
+    )
+    response = client.put_user_policy(PolicyDocument=policy_document_allow,
+                                      PolicyName='DenyPutPolicy', UserName=get_alt_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    # Copy object user1buck/srcobj to user2buck/dstobj  as user2
```
Comment (apoorva-rao) [inline s3tests_boto3/functional/test_iam.py]: Let's correct this comment here

**Review by apoorva-rao — CHANGES_REQUESTED** [review]

File: s3tests_boto3/functional/test_iam.py (hunk 5 above)
Comment (ketanarlulkar) [reply s3tests_boto3/functional/test_iam.py:1775]: The same policy has two statements to allow Put on user2buck and allow get on user1buck

File: s3tests_boto3/functional/test_iam.py (hunk 1 above)
Comment (ketanarlulkar) [reply s3tests_boto3/functional/test_iam.py:1637]: Now reading Etag of original object and comparing.

Comment (ketanarlulkar) [reply s3tests_boto3/functional/test_iam.py:1637]: Now giving permissions.
Updated TDS for head-object to get ETag

**Review by ravindrachoudhari — APPROVED** [review]

File: s3tests_boto3/functional/test_iam.py (hunk 5 above)
Comment (archanabpatil) [reply s3tests_boto3/functional/test_iam.py:1775]: Thanks for correcting TDS

**Review by archanabpatil — APPROVED** [review]

**Review by apoorva-rao — APPROVED** [review]

**Pull request merged**